File: CIVITASRUBY/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador_especulador < Jugador
    @@Factor_especulador = 2
    def self.nuevo_especulador(jugador, fianza)
      
      jugador = super(jugador)
      jugador.set_fianza(fianza)
      
      return jugador
    
    end

   
    def set_fianza(fianza)
      @fianza = fianza
    end
    def get_casas_max()
      return @@Casas_max * @@Factor_especulador
    end
    
    def get_hoteles_max()
      return @@Hoteles_max  * @@Factor_especulador
    end

    
    def debe_ser_encarcelado()
      if(self.encarcelado)
        return false
      else
        if puedo_gastar(@fianza)
          modificar_saldo(-@fianza)
          Diario.instance.ocurre_evento("El jugador " + "#{self.nombre}" + " se libra de la carcel porque ha pagado la fianza" )
          return false
        else if self.tiene_salvoconducto == false
            return true
        else
            perder_salvoconducto()
            Diario.instance.ocurre_evento("El jugador " + "#{self.nombre}" + " se libra de la carcel porque tenia el salvoconducto" )
            return false
        end
        
        end
      end
    end
    
    
   def paga_impuesto(cantidad)
     if self.encarcelado == true
       false
     else
       return paga(cantidad/@@Factor_especulador)
       
       
     end
   end
   
   def to_s
     propiedades = ""
       if @propiedades.length != 0
        @propiedades.each do |prop|
        propiedades += prop.to_s
        
        end
       else
         propiedades += " Aun no tiene propiedades"
        
      end
      
     resultado = "Jugador: Encarcelado = #{self.encarcelado}\n  Nombre = #{self.nombre} ESPECULADOR\n Saldo= #{self.saldo}
      \n Salvoconduct0 = #{self.salvoconducto}\n CasillaActual = #{self.num_casilla_actual}\n"
       
      resultado += "\nPropiedades = "+ propiedades
      return resultado
   end
    
  end
end

File: CIVITAS/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author Usuario
 */

import java.util.ArrayList;
        
public class JugadorEspeculador extends Jugador{
    
    protected static int FactorEspeculador = 2;
    private final int fianza;
    
    JugadorEspeculador(Jugador otro, int fianza){
        super(otro);
        actualizaPropietarioPorConversion();
        this.fianza = fianza;
    }
    
    
    
    private void actualizaPropietarioPorConversion(){
        ArrayList<TituloPropiedad> p = this.getPropiedades();
        
        for( TituloPropiedad t : p ){
            t.actualizaPropietarioPorConversion(this);
        }
    }
    
    @Override
    int getCasasMax(){return CasasMax * FactorEspeculador;}
    
    @Override 
    int getHotelesMax(){return HotelesMax * FactorEspeculador;}
    
    @Override
      protected Boolean debeSerEncarcelado(){
        if(this.encarcelado)
            return false;
        else{
            if(this.puedoGastar(fianza)){
                this.modificarSaldo(-fianza);
                Diario.getInstance().ocurreEvento("El jugador "+ this.getNombre() + " se libra de la carcel porque ha pagado la fianza");
                return false;
            }else if(!this.tieneSalvoconducto())
                return true;
            else{
                this.perderSalvoconducto();
                Diario.getInstance().ocurreEvento("El jugador "+ this.getNombre() + " se libra de la carcel con el salvoconducto");
                return false;
            }
        }
    }
    
    @Override
    Boolean pagaImpuesto(float cantidad){
        if(this.encarcelado)
            return false;
        else{
            return this.paga(cantidad / FactorEspeculador);
        }
    }
    
    @Override
    public String toString() {
        String result = "Jugador{" + " \nnombre=" + this.getNombre() + " ESPECULADOR" + "\nfianza="+ this.fianza + "\nencarcelado=" + encarcelado  + ", \nnumCasillaActual=" + this.getNumCasillaActual() + ", \npuedeComprar=" + this.getPuedeComprar() + ", \nsaldo=" + this.getSaldo() + ", ";
    
        if(this.getPropiedades() != null){
            for(TituloPropiedad t : this.getPropiedades()){
                result += "\n" + t.toString() + "\n";
            }
        }
        
        result += "\n}\n";
        
        return result;
    }
}

